js_modules/dagster-ui/packages/ui-core/src/runs/HourlyDataCache/HourlyDataCache.tsx

```diff
@@ -0,0 +1,266 @@
+type TimeWindow<T> = {start: number; end: number; data: T[]};
+
+export const ONE_HOUR_S = 60 * 60;
+
+type Subscription<T> = (data: T[]) => void;
+
+export class HourlyDataCache<T> {
+  private cache: Map<number, Array<TimeWindow<T>>> = new Map();
+  private subscriptions: Array<{hour: number; callback: Subscription<T>}> = [];
+
+  /**
+   * Adds data to the cache for the specified time range.
+   * @param start - The start time in seconds.
+   * @param end - The end time in seconds.
+   * @param data - The data to cache.
+   */
+  addData(start: number, end: number, data: T[]): void {
+    const startHour = Math.floor(start / ONE_HOUR_S);
+    const endHour = Math.floor(end / ONE_HOUR_S);
+
+    if (endHour - startHour > 1) {
+      throw new Error('Expected all data to fit within an hour');
+    }
+
+    this.addPartialData(startHour, start, end, data);
+    this.mergeCompleteHours();
+    this.notifySubscribers(startHour);
+  }
+
+  /**
+   * Adds data to the partial cache for a specific hour.
+   * @param hour - The hour for which to add data.
+   * @param start - The start time in seconds.
+   * @param end - The end time in seconds.
+   * @param data - The data to cache.
+   */
+  private addPartialData(hour: number, start: number, end: number, data: T[]): void {
+    if (!this.cache.has(hour)) {
+      this.cache.set(hour, []);
+    }
+    this.cache.get(hour)!.push({start, end, data});
+    this.cache.set(hour, this.mergeIntervals(this.cache.get(hour)!));
+  }
+
+  /**
+   * Merges complete hours in the cache.
+   */
+  private mergeCompleteHours(): void {
+    const completedHours: number[] = [];
+    for (const [hour, intervals] of this.cache) {
+      if (intervals.length === 1 && intervals[0]!.end - intervals[0]!.start === ONE_HOUR_S) {
+        completedHours.push(hour);
+      }
+    }
+    for (const hour of completedHours) {
+      this.cache.set(hour, [
+        {
+          start: hour * ONE_HOUR_S,
+          end: (hour + 1) * ONE_HOUR_S,
+          data: this.cache.get(hour)![0]!.data,
+        },
+      ]);
+    }
+  }
```

Comment:
this is rebuilding the entire cache every time? this function is confusing me a bit - maybe adding some comments would help

Reply:
Ooh I actually dont need this function, this was used in a previous incarnation of the cache and I forgot to remove it 😓